### PR TITLE
Add groups to env vars

### LINKED ISSUE #4014: declutter and organise ENV variables based settings
as of now the default variables are defined inside a big dict here `xonsh/environ.py:632`. It doesn't do any organisation of the variables. 

I want to organise them into groups like done in https://takoshell.org/documentation.html#section_27 . We can create a custom class (Settings) that gets value from env variables or return a default value. This singletons can be passed around/imported.

It can still work with existing code-base. We can also update the [documentation](https://xon.sh/envvars.html) page to show the variables under sub-headings. 

What do you think ?

### PR DESCRIPTION
**Pull request by jnoortheen**
fixes #4014

ToDos:
- [x] create groups for envvars
- [x] update document generated for envvars page

### CODE REVIEW

**Comment by jnoortheen** [comment]
env-vars-doc
here is a screen shot.

**Comment by jnoortheen** [comment]
any stoppers on this? I want to add some more environment variables to the config to customize sqlite-history backend to speedup startup.

**Comment by gforsyth** [comment]
Hey @jnoortheen -- I'll try to take a pass over this this week -- before I do, it looks like there are a bunch of empty artifacts in the env docs screenshot you posted.  Were those resolved in a later commit?

**Comment by jnoortheen** [comment]
the empty part was because of the browser, because of the large size.

File: xonsh/environ.py [hunk 1]
```diff
@@ -629,1152 +629,1038 @@ def default_lscolors(env):
...
+        "``XONSH_SHOW_TRACEBACK`` has been set. Its value must be a writable file "
+        "or None / the empty string if traceback logging is not desired. "
+        "Logging to a file is not enabled by default.",
+    )
+    STAR_PATH = Var.no_default("env_path", pattern=re.compile(r"\w*PATH$"))
+    STAR_DIRS = Var.no_default("env_path", pattern=re.compile(r"\w*DIRS$"))
+
+
+class ChangeDirSetting(Xettings):
+    """``cd`` Behaviour"""
```
Comment (gforsyth) [inline xonsh/environ.py:1074]:
```suggestion
    """``cd`` Behavior"""
```

We use US spelling in the rest of the docs, so I'd like to keep that consistent

File: xonsh/environ.py [hunk 2]
```diff
@@ -629,1152 +629,1038 @@ def default_lscolors(env):
...
+        "with Bash, xonsh always prefer an existing relative path.",
+        type_str="env_path",
+    )
+    DIRSTACK_SIZE = Var.with_default(
+        20,
+        "Maximum size of the directory stack.",
+    )
+    PUSHD_MINUS = Var.with_default(
+        False,
+        "Flag for directory pushing functionality. False is the normal " "behavior.",
```
Comment (gforsyth) [inline xonsh/environ.py:1097]:
```suggestion
        "Flag for directory pushing functionality. False is the normal behavior.",
```

File: xonsh/environ.py [hunk 3]
```diff
@@ -629,1152 +629,1038 @@ def default_lscolors(env):
...
+    )
+    PUSHD_SILENT = Var.with_default(
+        False,
+        "Whether or not to suppress directory stack manipulation output.",
+    )
+
+
+class InterpreterSetting(Xettings):
+    """Interpreter Behaviour"""
```
Comment (gforsyth) [inline xonsh/environ.py:1106]:
```suggestion
    """Interpreter Behavior"""
```

File: xonsh/environ.py [hunk 4]
```diff
@@ -629,1152 +629,1038 @@ def default_lscolors(env):
...
+        "A format string, using the same keys and colors as ``$PROMPT``, that "
+        "is appended whenever stderr is displayed. This may be used in "
+        "conjunction with ``$XONSH_STDERR_PREFIX`` to start the block."
+        "For example, to have stderr appear on a red background, the "
+        'prefix & postfix pair would be "{BACKGROUND_RED}" & "{RESET}".',
+    )
+
+
+class PTKSetting(PromptSetting):  # sub-classing -> sub-group
+    """Prompt Tool Kit shell
```
Comment (gforsyth) [inline xonsh/environ.py:1433]:
```suggestion
    """Prompt Toolkit shell
```

File: xonsh/environ.py [hunk 5]
```diff
@@ -629,1152 +629,1038 @@ def default_lscolors(env):
...
+    )
+    ENABLE_ASYNC_PROMPT = Var.with_default(
+        False,
+        "When enabled the prompt is rendered using threads. "
+        "$PROMPT_FIELD that take long will be updated in the background and will not affect prompt speed. ",
+    )
+
+
+class AutoCompletionSetting(Xettings):
+    """Tab-completion behaviour."""
```
Comment (gforsyth) [inline xonsh/environ.py:1510]:
```suggestion
    """Tab-completion behavior."""
```

File: xonsh/environ.py [hunk 6]
```diff
@@ -629,1152 +629,1038 @@ def default_lscolors(env):
...
+    )
+    SUBSEQUENCE_PATH_COMPLETION = Var.with_default(
+        True,
+        "Toggles subsequence matching of paths for tab completion. "
+        "If ``True``, then, e.g., ``~/u/ro`` can match ``~/lou/carcolh``.",
+    )
+
+
+class PTKCompletionSetting(AutoCompletionSetting):
+    """Prompt Tool Kit  auto-completion"""
```
Comment (gforsyth) [inline xonsh/environ.py:1565]:
```suggestion
    """Prompt Toolkit tab-completion"""
```

**Review by gforsyth — CHANGES_REQUESTED** [review]
Overall looks good, some nits on spelling

File: docs/jinja_rst_ext.py [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+            if file_path.exists():
+                ctx["content"] = environment.from_string(file_path.read_text()).render(
+                    **ctx
+                )
+
+        rendered = app.builder.templates.render_string(src, ctx)
+        source[0] = rendered
+
+        # for debugging purpose write output
+        Path(cur_dir / "_build" / f"{docname}.rst.out").write_text(rendered)
```
[the remark this answers is not in the document]
Comment (jnoortheen) [reply docs/jinja_rst_ext.py:40]: that was for debugging, seems I missed it.

File: docs/jinja_rst_ext.py [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    Render our pages as a jinja template for fancy templating goodness.
+    """
+    # Make sure we're outputting HTML
+    if app.builder.format != "html":
+        return
+
+    ctx = app.config.jinja_contexts.get(docname)
+    if ctx is not None:
+        environment = jinja2.Environment(
+            # loader=jinja2.FileSystemLoader("."),
```
[the remark this answers is not in the document]
Comment (jnoortheen) [reply docs/jinja_rst_ext.py]: done

File: docs/jinja_rst_ext.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (jnoortheen) [reply docs/jinja_rst_ext.py]: I've commented them.

**Comment by jnoortheen** [comment]
I tried to publish it to github pages here https://jnoortheen.github.io/xonsh/envvars.html
the css files aren't loaded. But you can have an overview from the image and this page of what will change.

**Comment by gforsyth** [comment]
Thanks for putting this in @jnoortheen !

**Pull request merged**